streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -16,78 +16,408 @@
+            Collectors.toMap(state -> state.processId().id(), KafkaStreamsState::rackId));
+
+        final Map<UUID, Set<TaskId>> taskIdsByProcess = clients.values().stream().collect(
+            Collectors.toMap(state -> state.processId().id(), state -> {
+                if (isStandby) {
+                    return state.previousStandbyTasks();
+                }
+                return state.previousActiveTasks();
+            })
+        );
+
+        final RackAwareGraphConstructor<UUID> graphConstructor = new MinTrafficGraphConstructor<>();
+        final AssignmentGraph assignmentGraph = buildTaskGraph(clientIds, clientRacks, taskIds, taskIdsByProcess, topicPartitionsByTaskId,
+            crossRackTrafficCost, nonOverlapCost, hasReplica, isStandby, graphConstructor);
+        return assignmentGraph.graph.totalCost();
+    }
+
+    private static AssignmentGraph buildTaskGraph(final List<UUID> clientIds,
+                                                  final Map<UUID, Optional<String>> clientRacks,
+                                                  final List<TaskId> taskIds,
+                                                  final Map<UUID, Set<TaskId>> previousTaskIdsByProcess,
+                                                  final Map<TaskId, Set<TaskTopicPartition>> topicPartitionsByTaskId,
+                                                  final int crossRackTrafficCost,
+                                                  final int nonOverlapCost,
+                                                  final boolean hasReplica,
+                                                  final boolean isStandby,
+                                                  final RackAwareGraphConstructor<UUID> graphConstructor) {
+        final Map<UUID, UUID> clientsUuidByUuid = clientIds.stream().collect(Collectors.toMap(id -> id, id -> id));
+        final Map<TaskId, UUID> clientByTask = new HashMap<>();
+        final Map<UUID, Integer> taskCountByClient = new HashMap<>();
+        final Graph<Integer> graph = graphConstructor.constructTaskGraph(
+            clientIds,
+            taskIds,
+            clientsUuidByUuid,
+            clientByTask,
+            taskCountByClient,
+            (processId, taskId) -> {
+                return previousTaskIdsByProcess.get(processId).contains(taskId);
+            },
+            (taskId, processId, inCurrentAssignment, unused0, unused1, unused2) -> {
+                final int assignmentChangeCost = !inCurrentAssignment ? nonOverlapCost : 0;
+                final Optional<String> clientRack = clientRacks.get(processId);
+                final Set<TaskTopicPartition> topicPartitions = topicPartitionsByTaskId.get(taskId).stream().filter(tp -> {
+                    return isStandby ? tp.isChangelog() : true;
+                }).collect(Collectors.toSet());
+                return assignmentChangeCost + getCrossRackTrafficCost(topicPartitions, clientRack, crossRackTrafficCost);
+            },
```

Comment:
We can't really have a simple this##getCost function because we have no `this`, which means we cannot easily move around the `clientRacks`, or `topicPartitionsByTaskId`. It's going to have to be a function pointer no matter how we spin it unfortunately.